core/node/state_keeper/src/batch_executor/main_executor.rs

```diff
@@ -205,7 +205,7 @@ impl CommandReceiver {
         }
 
         let tx_metrics = ExecutionMetricsForCriteria::new(Some(tx), &tx_result);
-        let gas_remaining = vm.gas_remaining();
+        let gas_remaining = tx_result.statistics.gas_remaining;
```

Comment:
is it equivalent?

Reply:
Do you mean in theory, or in practice? AFAICT, semantically, it's the same value, and it's set correctly for all VMs other than M5, M6 and 1.3.2 (e.g., [here's the latest VM](https://github.com/matter-labs/zksync-era/blob/8ed086afecfcad30bfda44fc4d29a00beea71cca/core/lib/multivm/src/versions/vm_latest/implementation/statistics.rs#L34), where `gas_remaining_after` is obtained [here](https://github.com/matter-labs/zksync-era/blob/8ed086afecfcad30bfda44fc4d29a00beea71cca/core/lib/multivm/src/versions/vm_latest/implementation/execution.rs#L76)); but these 3 can be fixed to return a correct value as well. An alternative would be to add a `gas_remaining` field to `VmExecutionResultAndLogs`, but it looks obviously redundant there since it includes statistics as well. Maybe, I'm misinterpreting `VmExecutionStatistics` (like, are its fields approximations / filled at best-effort basis)?

Reply:
I've added remaining gas tracking (I think correctly) for these 3 old VMs [in this commit](https://github.com/matter-labs/zksync-era/pull/2760/commits/056312f37c5e541098ccefbb8a234f0463cb42bd).